### PR TITLE
Add implementation of hierarchical poisson factorization(HPF) 

### PR DESCRIPTION
**Pull request by KimuraTian**
TODO:

1. Optimize the HPF model provider.
2. Come up with the way to construct artificial ratings with distributions of Figure 2 in the paper, and test this implementation.
3. Build more test and comments.

### CODE REVIEW

File: lenskit-pf/src/main/java/org/lenskit/pf/PMFModel.java [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+        rows = new Int2ObjectOpenHashMap<>();
+        gammaOrLambdaRteSecondTerm = new Int2DoubleOpenHashMap();
+    }
+
+    public PMFModel addEntry(ModelEntry entry) {
+        int row = entry.getRowNumber();
+        rows.put(row, entry);
+        int featureCount = entry.getFeatureCount();
+        for (int k = 0; k < featureCount; k++) {
+            double value = gammaOrLambdaRteSecondTerm.getOrDefault(k, 0) + entry.getGammaOrLambdaShpEntry(k) / entry.getGammaOrLambdaRteEntry(k);
```
Comment (mdekstrand) [inline lenskit-pf/src/main/java/org/lenskit/pf/PMFModel.java]: Use the `addTo` method of `Int2DoubleOpenHashMap` to reduce boxing and hashing

File: lenskit-pf/src/main/java/org/lenskit/pf/IsProbabilityPrediciton.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+/**
+ * Select either using poisson probability as prediction or using rate of poisson distribution as prediction
+ */
+@Documented
+@DefaultBoolean(true)
+@Parameter(Boolean.class)
+@Qualifier
+@Target({ElementType.METHOD, ElementType.PARAMETER})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface IsProbabilityPrediciton {
```
Comment (mdekstrand) [inline lenskit-pf/src/main/java/org/lenskit/pf/IsProbabilityPrediciton.java]: `IsProbabilityPrediction`

File: lenskit-pf/src/main/java/org/lenskit/pf/ItemActivityPriorMean.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import javax.inject.Qualifier;
+import java.lang.annotation.*;
+
+@Documented
+@DefaultDouble(1.0)
+@Parameter(Double.class)
+@Qualifier
+@Target({ElementType.METHOD, ElementType.PARAMETER})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface ItemActivityPriorMean {
```
Comment (mdekstrand) [inline lenskit-pf/src/main/java/org/lenskit/pf/ItemActivityPriorMean.java]: We need JavaDoc explaining this.

File: lenskit-pf/src/main/java/org/lenskit/pf/ItemActivityShpPrior.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+import javax.inject.Qualifier;
+import java.lang.annotation.*;
+
+@Documented
+@DefaultDouble(0.3)
+@Parameter(Double.class)
+@Qualifier
+@Target({ElementType.METHOD, ElementType.PARAMETER})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface ItemActivityShpPrior {
```
Comment (mdekstrand) [inline lenskit-pf/src/main/java/org/lenskit/pf/ItemActivityShpPrior.java]: We need JavaDoc explaining this. Also, should we call it the 'shape prior' or the 'prior shape'?

File: lenskit-pf/src/main/java/org/lenskit/pf/ItemWeightShpPrior.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+import javax.inject.Qualifier;
+import java.lang.annotation.*;
+
+@Documented
+@DefaultDouble(0.3)
+@Parameter(Double.class)
+@Qualifier
+@Target({ElementType.METHOD, ElementType.PARAMETER})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface ItemWeightShpPrior {
```
Comment (mdekstrand) [inline lenskit-pf/src/main/java/org/lenskit/pf/ItemWeightShpPrior.java]: Comment + consider name.

File: lenskit-pf/src/main/java/org/lenskit/pf/MaxRandomOffsetForShape.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+import org.lenskit.inject.Parameter;
+
+import javax.inject.Qualifier;
+import java.lang.annotation.*;
+
+/**
+ * Max value of the random offset when initializing the shape parameters
+ */
+@Documented
+@DefaultDouble(0.01)
```
Comment (mdekstrand) [inline lenskit-pf/src/main/java/org/lenskit/pf/MaxRandomOffsetForShape.java]: That looks pretty small!

File: lenskit-pf/src/main/java/org/lenskit/pf/PMFModel.java [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+import org.apache.commons.math3.linear.MatrixUtils;
+import org.apache.commons.math3.linear.RealVector;
+import org.apache.commons.math3.special.Gamma;
+import org.lenskit.data.ratings.RatingMatrixEntry;
+
+import java.util.*;
+import java.util.stream.IntStream;
+
+
+public class PMFModel {
```
Comment (mdekstrand) [inline lenskit-pf/src/main/java/org/lenskit/pf/PMFModel.java]: This is just the internal class, right? Let's give it a clearer name and make it package-private (remove `public`).

File: lenskit-pf/src/main/java/org/lenskit/pf/PMFModelCollector.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+import java.util.EnumSet;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.BinaryOperator;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collector;
+
+
+public class PMFModelCollector implements Collector<PMFModel.ModelEntry, PMFModel, PMFModel> {
```
Comment (mdekstrand) [inline lenskit-pf/src/main/java/org/lenskit/pf/PMFModelCollector.java]: This should probably be package-private.

File: lenskit-pf/src/main/java/org/lenskit/pf/RandomDataSplitStrategyProvider.java [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+        final int userNum = snapshot.userIndex().size();
+        final int itemNum = snapshot.itemIndex().size();
+        logger.info("Rating matrix size: {} users and {} items", userNum, itemNum);
+        List<RatingMatrixEntry> allRatings = new ArrayList<>(snapshot.getRatings());
+        final int size = allRatings.size();
+        final int validationSize = Math.toIntExact(Math.round(size*proportion));
+        logger.info("validation set size: {} ratings", validationSize);
+        Collections.shuffle(allRatings, random);
+        List<RatingMatrixEntry> subList = allRatings.subList(0, validationSize);
+//        IntSet randomIndices = new IntOpenHashSet();
```
Comment (mdekstrand) [inline lenskit-pf/src/main/java/org/lenskit/pf/RandomDataSplitStrategyProvider.java]: Let's delete unused code like this.

File: lenskit-pf/src/main/java/org/lenskit/pf/RandomSeed.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+import org.lenskit.inject.Parameter;
+
+import javax.inject.Qualifier;
+import java.lang.annotation.*;
+
+/**
+ * Seed of random generator used by {@link RandomDataSplitStrategyProvider}
+ */
+@Documented
+@DefaultInteger(0)
```
Comment (mdekstrand) [inline lenskit-pf/src/main/java/org/lenskit/pf/RandomSeed.java]: We should do random seed differently, but we should do the whole random split differently. Leave it as-is for now.

File: lenskit-pf/src/main/java/org/lenskit/pf/UserActivityShpPrior.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+import javax.inject.Qualifier;
+import java.lang.annotation.*;
+
+@Documented
+@DefaultDouble(0.3)
+@Parameter(Double.class)
+@Qualifier
+@Target({ElementType.METHOD, ElementType.PARAMETER})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface UserActivityShpPrior {
```
Comment (mdekstrand) [inline lenskit-pf/src/main/java/org/lenskit/pf/UserActivityShpPrior.java]: Same comment as for item activity.

File: lenskit-pf/src/main/java/org/lenskit/pf/UserWeightShpPrior.java [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+import javax.inject.Qualifier;
+import java.lang.annotation.*;
+
+@Documented
+@DefaultDouble(0.3)
+@Parameter(Double.class)
+@Qualifier
+@Target({ElementType.METHOD, ElementType.PARAMETER})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface UserWeightShpPrior {
```
Comment (mdekstrand) [inline lenskit-pf/src/main/java/org/lenskit/pf/UserWeightShpPrior.java]: Same as for other shape priors.

File: lenskit-pf/src/test/java/org/lenskit/pf/RandomDataSplitStrategyProviderTest.java [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+                lambdaRte.setEntry(i, k, valueRte);
+            }
+            double tShp = aPrime + maxOffsetShp*random1.nextDouble();
+            tauRte.setEntry(i, tRte);
+            tauShp.setEntry(i, tShp);
+        }
+        timer.stop();
+        System.out.println("time for sequential is " + timer.getTime());
+
+//        Int2ObjectMap userModel = preUserModel.getModel();
```
Comment (mdekstrand) [inline lenskit-pf/src/test/java/org/lenskit/pf/RandomDataSplitStrategyProviderTest.java]: Let's delete unused code.

File: lenskit-pf/src/main/java/org/lenskit/pf/AbsErrorStoppingCondition.java [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+        return new Controller();
+    }
+
+
+    private class Controller implements TrainingLoopController {
+        private int iterations = 0;
+
+        @Override
+        public boolean keepTraining(double error) {
+            if (error < threshold || iterations >= maxIterCount) {
```
[the remark this answers is not in the document]
Comment (KimuraTian) [reply lenskit-pf/src/main/java/org/lenskit/pf/AbsErrorStoppingCondition.java]: Removed this stopping Condition

File: lenskit-pf/src/main/java/org/lenskit/pf/ConvergenceCheckFrequency.java [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+package org.lenskit.pf;
+
+import org.grouplens.grapht.annotation.DefaultInteger;
+import org.lenskit.inject.Parameter;
+
+import javax.inject.Qualifier;
+import java.lang.annotation.*;
+
+/**
+ * The number of frequency used to compute the convergence condition (average predictive log likelihood of validation ratings)
```
[the remark this answers is not in the document]
Comment (KimuraTian) [reply lenskit-pf/src/main/java/org/lenskit/pf/ConvergenceCheckFrequency.java]: The convergence is based on the relative change of previous calculation and current calculation, and the calculation is the average of predictive log likelihood values over all validation ratings. So this is for reducing some training time for large datasets.

File: lenskit-pf/src/main/java/org/lenskit/pf/HPFItemScorer.java [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+
+public class HPFItemScorer extends AbstractItemScorer {
+    private static Logger logger = LoggerFactory.getLogger(HPFItemScorer.class);
+
+    private final HPFModel model;
+    private final boolean isProbPredition;
```
[the remark this answers is not in the document]
Comment (KimuraTian) [reply lenskit-pf/src/main/java/org/lenskit/pf/HPFItemScorer.java]: Corrected. Thanks.

File: lenskit-pf/src/main/java/org/lenskit/pf/ItemActivityPriorMean.java (hunk 3 above)
Comment (KimuraTian) [reply lenskit-pf/src/main/java/org/lenskit/pf/ItemActivityPriorMean.java]: Added some simple explanations.

File: lenskit-pf/src/main/java/org/lenskit/pf/PMFModel.java (hunk 7 above)
Comment (KimuraTian) [reply lenskit-pf/src/main/java/org/lenskit/pf/PMFModel.java]: Removed.

**Pull request merged**